extensions/amp-dailymotion/1.0/component.js

```diff
@@ -0,0 +1,132 @@
+/**
+ * Copyright 2021 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as Preact from '#preact';
+import {forwardRef} from '#preact/compat';
+import {
+  useCallback,
+  useEffect,
+  useLayoutEffect,
+  useMemo,
+  useRef,
+  useState,
+} from '#preact';
+import {dispatchCustomEvent} from '#core/dom';
+import {
+  DailymotionEvents,
+  getDailymotionIframeSrc,
+  makeDailymotionMessage,
+} from '../dailymotion-api';
+import {VideoIframe} from '../../amp-video/1.0/video-iframe';
+import {parseQueryString} from '#core/types/string/url';
+
+/**
+ * @param {string} method
+ * @return {string}
+ */
+function makeMethodMessage(method) {
+  switch (method) {
+    case 'mute':
+      return makeDailymotionMessage('muted', [true]);
+      break;
+    case 'unmute':
+      return makeDailymotionMessage('muted', [false]);
+      break;
+    case 'showControls':
+      return makeDailymotionMessage('controls', [true]);
+      break;
+    case 'hideControls':
+      return makeDailymotionMessage('controls', [false]);
+      break;
+
+    default:
+      return makeDailymotionMessage(method);
+  }
+}
+
+/**
+ * @param {MessageEvent} event
+ */
+function onMessage({currentTarget, data}) {
+  const parsedData = parseQueryString(/** @type {string} */ (data));
+  const event = parsedData?.['event'];
+  if (event === DailymotionEvents.API_READY) {
+    dispatchCustomEvent(currentTarget, 'canplay');
+    return;
+  }
+  if (DailymotionEvents[event]) {
+    dispatchEvent(currentTarget, DailymotionEvents.event);
+    return;
+  }
+}
+
+/**
+ * @param {!DailymotionDef.Props} props
+ * @param ref
```

Comment:
This can be either the video API or a defined Dailymotion API:
```suggestion
 * @param {{current: ?VideoWrapperDef.Api}} ref
```
```suggestion
 * @param {{current: ?DailymotionDef.DailymotionApi}} ref
```

Reply:
Is there a reason to use one over the other?

Reply:
Meaning wise, the first one is already defined and can be used if the API is identical. The second one we would use if the API is different from the other and needs to be separately defined. cc/ @samouri in case type checking changes-to-come has another answer here.